### PR TITLE
Bug Fixes and Reorder arrays

### PR DESCRIPTION
**Pull request by qmacAstanford**
Reorder R(NT,3) to R(3,NT) because I usually access x,y,z in succession.
Also make several Bug Fixes that probably only affect me.

### CODE REVIEW

File: src/mc/MC_confine.f03 [hunk 1]
```diff
@@ -32,42 +31,42 @@ subroutine MC_confine(confineType, LBox, RP, NT, IT1, IT2, ECon)
...
+if (confineType == 'none') then
     return
-elseif(confineType == 1) then
+elseif(confineType == 'platesInZ') then
     ! Confinement only in the z-direction
     ! limits: 0 and LBox
     do I = IT1,IT2
-        if(RP(I,3)<0.0_dp) then
+        if(RP(3,I)<0.0_dp) then
             ECon = 9990000.0_dp
```
Comment (brunobeltran) [inline src/mc/MC_confine.f03]: We should make this large energy a constant. And while we're at it, we should do this for all "just put a big energy here" numbers.

File: src/mc/MCsim.f03 [hunk 2]
```diff
@@ -53,6 +53,7 @@ subroutine MCsim(wlc_p,wlc_d,NSTEP)
     Type(wlcsim_data), intent(inout) :: wlc_d     ! system allocated data
     integer DELTA             !Alexander polynomial evaluated at t = -1; used for knot checking
     real(dp) para(10)
+    integer m_plus3
```
Comment (brunobeltran) [inline src/mc/MCsim.f03]: comment that this "m" is a spherical harmonic quantum "m"

File: src/mc/MC_int_swap.f03 [hunk 3]
```diff
@@ -29,11 +30,13 @@ subroutine MC_int_swap(wlc_p,wlc_d,I1,I2,I3,I4)
 real(dp) RBin(3)    ! bead position
 integer inDBin              ! index of bin
 integer ISX,ISY,ISZ
-LOGICAL isA   ! The bead is of type A
-
-! Copy so I don't have to type wlc_p% everywhere
+!LOGICAL isA   ! The bead is of type A
+integer AminusB
+integer m_plus3
```
Comment (brunobeltran) [inline src/mc/MC_int_swap.f03:35]: same

File: src/mc/MC_int_rep.f03 [hunk 4]
```diff
@@ -33,12 +33,12 @@ subroutine MC_int_rep(wlc_p,wlc_d,I1,I2,forward)
 integer inDBin              ! index of bin
 integer ISX,ISY,ISZ
 LOGICAL isA   ! The bead is of type A
-
-! Copy so I don't have to type wlc_p% everywhere
+real(dp) phi2(5)
+integer m_plus3
```
Comment (brunobeltran) [inline src/mc/MC_int_rep.f03]: same

File: src/mc/MC_int_chem.f03 [hunk 5]
```diff
@@ -0,0 +1,163 @@
...
+real(dp) RBin(3)    ! bead position
+integer inDBin              ! index of bin
+integer ISX,ISY,ISZ
+LOGICAL isA   ! The bead is of type A
+
+! Copy so I don't have to type wlc_p% everywhere
+integer NBinX(3)
+real(dp) temp
+real(dp) contribution 
+integer m_plus3
```
Comment (brunobeltran) [inline src/mc/MC_int_chem.f03]: same

File: src/mc/Y2calc.f03 [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+!
+!     Written by Quinn MacPherson 8/4/17
+!---------------------------------------------------------------!
+subroutine Y2calc(u,Y2real)
+use params, only: dp 
+implicit none
+
+real(dp), intent(in) :: u(3)
+real(dp), intent(out) :: Y2real(5)
+real(dp), parameter, dimension(5) :: coefficient = (/1.092548430592079, &
```
Comment (brunobeltran) [inline src/mc/Y2calc.f03]: replace with symbolic values

File: src/mc/verifyEnergiesFromScratch.f03 [hunk 7]
```diff
@@ -90,7 +90,7 @@ subroutine VerifyEnergiesFromScratch(wlc_p, wlc_d)
    call CalculateEnergiesFromScratch(wlc_p, wlc_d)
 
     ! --- Binding Energy ---
-    if(abs(wlc_d%EBind-wlc_d%DEBind).gt.0.00001) then
+    if(abs(wlc_d%EBind-wlc_d%DEBind) > 0.00001) then
```
Comment (brunobeltran) [inline src/mc/verifyEnergiesFromScratch.f03]: replace with hardcoded float eps in params

File: src/mc/verifyEnergiesFromScratch.f03 [hunk 8]
```diff
@@ -99,7 +99,7 @@ subroutine VerifyEnergiesFromScratch(wlc_p, wlc_d)
     wlc_d%x_mu = wlc_d%dx_mu
 
     ! --- Elastic Energy ---
-    if(abs((wlc_d%EElas(1)+  wlc_d%EElas(2)+ wlc_d%EElas(3))-&
+    if(abs((wlc_d%EElas(1) +  wlc_d%EElas(2) + wlc_d%EElas(3))-&
```
Comment (brunobeltran) [inline src/mc/verifyEnergiesFromScratch.f03:102]: replace with hardcoded float eps in params

File: src/mc/verifyEnergiesFromScratch.f03 [hunk 9]
```diff
@@ -112,34 +112,44 @@ subroutine VerifyEnergiesFromScratch(wlc_p, wlc_d)
     ! --- Interaction Energy ---
     if (wlc_p%field_int_on) then
         ! test to see if sum of changes are same as calculating from scratch
-        if(abs(wlc_d%EChi-wlc_d%DEChi).gt. 0.0001_dp) then
+        if(abs(wlc_d%EChi-wlc_d%DEChi) > 0.0001_dp) then
```
Comment (brunobeltran) [inline src/mc/verifyEnergiesFromScratch.f03]: replace with hardcoded float eps in params

File: src/mc/verifyEnergiesFromScratch.f03 [hunk 10]
```diff
@@ -112,34 +112,44 @@ subroutine VerifyEnergiesFromScratch(wlc_p, wlc_d)
...
              write(ERROR_UNIT,*) "Warning. Intigrated chi energy:", &
                      wlc_d%EChi,"  while absolute chi energy:", &
                      wlc_d%DEChi," save point mc_ind = ",wlc_d%mc_ind
         endif
         wlc_d%EChi = wlc_d%DEChi
         wlc_d%x_chi = wlc_d%dx_chi
-        if(abs(wlc_d%ECouple-wlc_d%DECouple).gt. 0.0001_dp) then
+        if(abs(wlc_d%ECouple-wlc_d%DECouple) > 0.0001_dp) then
```
Comment (brunobeltran) [inline src/mc/verifyEnergiesFromScratch.f03]: replace with hardcoded float eps in params

File: src/mc/verifyEnergiesFromScratch.f03 [hunk 11]
```diff
@@ -112,34 +112,44 @@ subroutine VerifyEnergiesFromScratch(wlc_p, wlc_d)
...
              write(ERROR_UNIT,*) "Warning. Intigrated couple energy:", &
                      wlc_d%ECouple,"  while absolute couple energy:", &
                      wlc_d%DECouple," save point mc_ind = ",wlc_d%mc_ind
         endif
         wlc_d%ECouple = wlc_d%DECouple
         wlc_d%x_Couple = wlc_d%dx_couple
-        if(abs(wlc_d%EKap-wlc_d%DEKap).gt. 0.0001_dp) then
+        if(abs(wlc_d%EKap-wlc_d%DEKap) > 0.0001_dp) then
```
Comment (brunobeltran) [inline src/mc/verifyEnergiesFromScratch.f03]: replace with hardcoded float eps in params

File: src/mc/verifyEnergiesFromScratch.f03 [hunk 12]
```diff
@@ -112,34 +112,44 @@ subroutine VerifyEnergiesFromScratch(wlc_p, wlc_d)
...
              write(ERROR_UNIT,*) "Warning. Intigrated Kap energy:", &
                      wlc_d%EKap,"  while absolute Kap energy:", &
                      wlc_d%DEKap," save point mc_ind = ",wlc_d%mc_ind
         endif
         wlc_d%EKap = wlc_d%DEKap
         wlc_d%x_Kap = wlc_d%dx_Kap
 
-        if(abs(wlc_d%EField-wlc_d%DEField).gt.0.00001) then
+        if(abs(wlc_d%EField-wlc_d%DEField) > 0.00001) then
```
Comment (brunobeltran) [inline src/mc/verifyEnergiesFromScratch.f03]: replace with hardcoded float eps in params

File: src/mc/verifyEnergiesFromScratch.f03 [hunk 13]
```diff
@@ -112,34 +112,44 @@ subroutine VerifyEnergiesFromScratch(wlc_p, wlc_d)
...
             write(ERROR_UNIT,*) "Warning. Integrated field enrgy:", &
                     wlc_d%EField," while absolute field energy:", &
                     wlc_d%DEField," save point mc_ind = ",wlc_d%mc_ind
         endif
         wlc_d%EField = wlc_d%DEField
         wlc_d%x_Field = wlc_d%dx_Field
+
+        if(wlc_p%chi_l2_on) then
+            if(abs(wlc_d%EMaiersaupe-wlc_d%deMaierSaupe) > 0.00001) then
```
Comment (brunobeltran) [inline src/mc/verifyEnergiesFromScratch.f03]: replace with hardcoded float eps in params

File: src/util/binning.f03 [hunk 14]
```diff
@@ -171,25 +171,25 @@ recursive subroutine addBead(bin,R,NT,beadID)
         implicit none
         Type(binType), intent(inout) :: bin
         integer, intent(in) :: NT ! total number of beads
-        real(dp), intent(in) :: R(NT,3)  ! locations of all beads
+        real(dp), intent(in) :: R(3,NT)  ! locations of all beads
         integer, intent(in) :: beadID
         integer binIndex
 
         real(dp), parameter :: eps = 0.00000001
```
Comment (brunobeltran) [inline src/util/binning.f03]: should inherit from params instead

File: src/util/getdim.f03 [hunk 15]
```diff
@@ -35,18 +35,18 @@ subroutine getdim(N,NP,NT,R,RCOM,RCOMSQ,RGYRSQ,R2,R4,R6,DR)
...
-     X(:,2) = R(IB1:IBN,2)
-     X(:,3) = R(IB1:IBN,3)
-     RCOM(IP,:) = SUM(X,DIM = 1)/REAL(N)
-     XGYR(:,1) = X(:,1)-RCOM(IP,1)
-     XGYR(:,2) = X(:,2)-RCOM(IP,2)
-     XGYR(:,3) = X(:,3)-RCOM(IP,3)
+     X(1,:) = R(1,IB1:IBN)
+     X(2,:) = R(2,IB1:IBN)
+     X(3,:) = R(3,IB1:IBN)
+     RCOM(:,IP) = SUM(X,DIM = 1)/REAL(N)
```
Comment (brunobeltran) [inline src/util/getdim.f03]: dim should have changed

File: src/wlcsim/MCparrll_mpi.f90 [hunk 16]
```diff
@@ -183,6 +186,7 @@ subroutine replicaExchange(wlc_p, wlc_d)
     !wlc_p%para(1)  =cof(6)
     !wlc_p%para(2)  =cof(7)
     !wlc_p%para(3)  =cof(8)
+    wlc_p%chi_l2 = cof(9)
 
     if (abs(wlc_d%EChi-x(1)*CofOld(1)).gt.0.0000001_dp) then
```
Comment (brunobeltran) [inline src/wlcsim/MCparrll_mpi.f90]: immediately unpack arr and do things with names
for x and cof

Comment (brunobeltran) [inline src/wlcsim/MCparrll_mpi.f90]: params eps again

File: src/wlcsim/MCparrll_mpi.f90 [hunk 17]
```diff
@@ -198,6 +202,7 @@ subroutine replicaExchange(wlc_p, wlc_d)
    ! wlc_p%EElas(1) = wlc_p%EElas(1) + x(6)*(Cof(6)-CofOld(6))
    ! wlc_p%EElas(2) = wlc_p%EElas(2) + x(7)*(Cof(7)-CofOld(7))
    ! wlc_p%EElas(3) = wlc_p%EElas(3) + x(8)*(Cof(8)-CofOld(8))
+    wlc_d%eMaierSaupe    =wlc_d%eMaierSaupe    +x(9)*(Cof(5)-CofOld(5))
 
     if (abs(wlc_d%EChi-x(1)*Cof(1)).gt.0.000001_dp) then
```
Comment (brunobeltran) [inline src/wlcsim/MCparrll_mpi.f90]: params eps again

File: src/mc/MC_move.f90 [hunk 18]
```diff
@@ -203,50 +207,49 @@ subroutine MC_move(R,U,RP,UP,NT,NB,NP,IP,IB1,IB2,IT1,IT2,MCTYPE &
...
+        UP(3,I) = ROT(3,1)*U(1,I) + ROT(3,2)*U(2,I) + ROT(3,3)*U(3,I)
         I = I + 1
 
-
      ENDdo
 
   !  ------begining testing---------
   if(.false.) then
       ! This is a code block for testing
-      if (abs(RP(IT1,1)-R(IT1,1)).gt.0.000001) then
```
Comment (brunobeltran) [inline src/mc/MC_move.f90]: params eps

File: src/mc/MC_move.f90 [hunk 19]
```diff
@@ -203,50 +207,49 @@ subroutine MC_move(R,U,RP,UP,NT,NB,NP,IP,IB1,IB2,IT1,IT2,MCTYPE &
...
+      if (abs(RP(1,IT1)-R(1,IT1)).gt.0.000001) then
```
Comment (brunobeltran) [inline src/mc/MC_move.f90]: params eps

File: src/mc/MC_move.f90 [hunk 20]
```diff
@@ -203,50 +207,49 @@ subroutine MC_move(R,U,RP,UP,NT,NB,NP,IP,IB1,IB2,IT1,IT2,MCTYPE &
...
           print*, "error in crank-shaft move"
-          print*, RP(IT1,1), R(IT1,1)
+          print*, RP(1,IT1), R(1,IT1)
           stop 1
       endif
-      if (abs(RP(IT2,1)-R(IT2,1)).gt.0.000001) then
```
Comment (brunobeltran) [inline src/mc/MC_move.f90]: params eps

File: src/mc/MC_move.f90 [hunk 21]
```diff
@@ -203,50 +207,49 @@ subroutine MC_move(R,U,RP,UP,NT,NB,NP,IP,IB1,IB2,IT1,IT2,MCTYPE &
...
-          d2 = (RP(IT1 + 1,1)-RP(IT1,1))**2 + &
-             (RP(IT1 + 1,2)-RP(IT1,2))**2 + &
-             (RP(IT1 + 1,3)-RP(IT1,3))**2
+          d1 = (R(1,IT1 + 1)-R(1,IT1))**2 + &
+             (R(2,IT1 + 1)-R(2,IT1))**2 + &
+             (R(3,IT1 + 1)-R(3,IT1))**2
+          d2 = (RP(1,IT1 + 1)-RP(1,IT1))**2 + &
+             (RP(2,IT1 + 1)-RP(2,IT1))**2 + &
+             (RP(3,IT1 + 1)-RP(3,IT1))**2
           if (abs(d1-d2).gt.0.000001) then
```
Comment (brunobeltran) [inline src/mc/MC_move.f90]: params eps

File: src/mc/MC_move.f90 [hunk 22]
```diff
@@ -752,16 +754,16 @@ subroutine test_equiv_forward(U,R,UP,RP,NT,IT1,IT2,RparaMag,RperpMag)
...
+drOld(3) = R(3,IT1 + 1)-R(3,IT1)
+DRPAROld = DROld(1)*U(1,IT1) + DROld(2)*U(2,IT1) + DROld(3)*U(3,IT1)
+drNew(1) = RP(1,IT2)-RP(1,IT2-1)
+drNew(2) = RP(2,IT2)-RP(2,IT2-1)
+drNew(3) = RP(3,IT2)-RP(3,IT2-1)
+DRPARNew = DRNew(1)*UP(1,IT2-1) + &
+         DRNew(2)*UP(2,IT2-1) + &
+         DRNew(3)*UP(3,IT2-1)
 if (abs(drOld(1)**2 + drOld(2)**2 + drOld(3)**2&
       -(drNew(1)**2 + drNew(2)**2 + drNew(3)**2)).gt.0.000001) then
```
Comment (brunobeltran) [inline src/mc/MC_move.f90]: params eps

File: src/mc/MC_move.f90 [hunk 23]
```diff
@@ -776,12 +778,12 @@ subroutine test_equiv_forward(U,R,UP,RP,NT,IT1,IT2,RparaMag,RperpMag)
...
-drPerpNew(3) = drNew(3)-drParNew*UP(IT2-1,3)
+drPerpOld(1) = drOld(1)-drParOld*U(1,IT1)
+drPerpOld(2) = drOld(2)-drParOld*U(2,IT1)
+drPerpOld(3) = drOld(3)-drParOld*U(3,IT1)
+drPerpNew(1) = drNew(1)-drParNew*UP(1,IT2-1)
+drPerpNew(2) = drNew(2)-drParNew*UP(2,IT2-1)
+drPerpNew(3) = drNew(3)-drParNew*UP(3,IT2-1)
 
 if (abs(drPerpOld(1)**2 + drPerpOld(2)**2 + drPerpOld(3)**2 &
       -(drPerpNew(1)**2 + drPerpNew(2)**2 + drPerpNew(3)**2)).gt.0.000001_dp) then
```
Comment (brunobeltran) [inline src/mc/MC_move.f90]: params eps

File: src/wlcsim/params.f03 [hunk 24]
```diff
@@ -347,27 +360,31 @@ subroutine set_param_defaults(wlc_p,wlc_d)
         ! energy parameters
         wlc_p%EPS =0.3_dp ! TOdo: not input
         wlc_p%CHI =0.0_dp ! don't use chi by default
+        wlc_p%CHI_l2 =0.0_dp ! don't use maier Saupe by default
         wlc_p%hA =0.0_dp  ! don't use weird artificial field by default
-        wlc_p%KAP =10.0_dp ! "fairly" incompressible --Quinn
+        wlc_p%KAP =0.0_dp ! "fairly" incompressible --Quinn
```
Comment (brunobeltran) [inline src/wlcsim/params.f03]: no longer "fairly" incompressible lol

File: src/wlcsim/params.f03 [hunk 25]
```diff
@@ -1220,7 +1260,7 @@ subroutine printDescription(wlc_p)
         print*, " L0 = ", wlc_p%L0
         print*, " volume fraction polymer =", wlc_p%Fpoly
         print*, " bead volume V = ", wlc_p%beadVolume
-        print*, " number of kuhn lengths between beads, eps ", wlc_p%eps
+        print*, " 1/number of kuhn lengths between beads, eps ", wlc_p%eps
```
Comment (brunobeltran) [inline src/wlcsim/params.f03]: nope, was correct as-was

File: src/wlcsim/params.f03 [hunk 26]
```diff
@@ -992,19 +1030,23 @@ subroutine initialize_wlcsim_data(wlc_d, wlc_p)
 #if MPI_VERSION
         call init_MPI(wlc_d)
 #endif
-        allocate(wlc_d%R(NT,3))
-        allocate(wlc_d%U(NT,3))
+        allocate(wlc_d%R(3,NT))
+        allocate(wlc_d%U(3,NT))
         if (wlc_p%codeName /= 'bruno' .OR. wlc_p%nInitMCSteps /= 0) then
-            allocate(wlc_d%RP(NT,3))
```
Comment (brunobeltran) [inline src/wlcsim/params.f03:998]: these can be pulled out of the if statement now actually, can we do that here? thanks!

**Review by brunobeltran — CHANGES_REQUESTED** [review]
recommend cof and cofMtrx renamed to coefs in src/wlcsim/MCparrll_mpi.f90 and src/wlcsim/wlcsim_quinn.f03.

all changes minor, except one bug in BD. 

Accepted for merge with these changes.

File: src/wlcsim/params.f03 (hunk 26 above)
Comment (qmacAstanford) [reply src/wlcsim/params.f03:998]: I always need RP but I don't know if you do when you are only doing bd.  I'll let you make that call.

File: src/mc/verifyEnergiesFromScratch.f03 (hunk 8 above)
Comment (qmacAstanford) [reply src/mc/verifyEnergiesFromScratch.f03:102]: fixed

File: src/mc/MC_int_swap.f03 (hunk 3 above)
Comment (qmacAstanford) [reply src/mc/MC_int_swap.f03:35]: fixed

File: src/mc/Y2calc.f03 (hunk 6 above)
Comment (qmacAstanford) [reply src/mc/Y2calc.f03]: fixed as much as I can.  Replaced with locally defined PI.  Can't used pi from params as it isn't a parameter

**Review by brunobeltran — APPROVED** [review]

**Comment by brunobeltran** [comment]
I am able to compile the code using params::pi in definition of constants. We should clear this up then merge these changes! Super great work with all this cleanup and new code!

**Pull request merged**